🚧 WIP: V2

GGGava opened the pull request:
This is draft PR for a complete overhaul of react-native-health library. This code is a work in progress and is subject to change at any moment. You are welcome to test it but this is **not** stable or production ready.

GGGava committed:
- Initial implementation of the swift core with quantity type samples

GGGava committed:
- Improvements on aggregation queries

GGGava committed:
- Added metadata field to samples
- Samples are now encodable to String for bridging to react native

GGGava committed:
- Fixed typo
- Moved queries to a different file

GGGava committed:
- Added save quantity sample method
- Fixed issue with metadata predicate

GGGava committed:
- Created inser opearation with metadata

GGGava committed:
- Recreated example app
- Deleted old rn bindings
- Created bindings for new core

GGGava committed:
- Fixed issue in initHealthKit wrapper

GGGava committed:
- Added getQuantitySamples to RN bridge

GGGava committed:
- Exported getQuantitySamplesAggregation method

GGGava committed:
- Exported saveQuantitySample method

GGGava committed:
- Added initHealthKit support for older iOS versions

ruan-azevedo committed:
Migrate to typescript

ruan-azevedo committed:
Fix linter errors on the example app

GGGava committed:
Merge branch 'swift' of github.com:agencyenterprise/react-native-health into swift

GGGava committed:
- Exported units and AggregationOptions

GGGava committed:
- Added typing to exported functions
- Moved sample type parameter to parameter object
- Interval and anchor date parameters are now available in react native

GGGava committed:
- Sample dates are now string

GGGava committed:
- Metadata is now visible in js
- Fixed typo
- Fixed issue when saving sample
- Added unit field to quantity sampl- Fixed typo
- Fixed issue when saving sample
- Added unit field to quantity sample

GGGava committed:
- Changed aggregation queries to statistics queries

GGGava committed:
- Added activityType and IDs filter to workout queries

LucasAssisRo commented on v2/RNHealthKitCore/HealthKitCore/HealthKitTypes.swift:
```diff
@@ -0,0 +1,143 @@
+import HealthKit
+
+public protocol HealthKitType: RawRepresentable {
+    var type: HKSampleType { get }
```
suggestion: Since this protocol is not directly available in objective-c maybe you could add this `HKSampleType` as a primary associated type to the protocol. So you don't need to cast the type when using a concrete type.

I totally understand not going for it if the compiler is giving you grief for using PATs, an alternative could be having a computed property that does the type casting on the concrete types. 
```swift
enum QuantityType {
/// Don't use this directly. Use `quantitySampleType` instead.
public var type: HKSampleType { ... } // I wonder if there's a way to use @available here to help 
public var quantitySampleType: HKQuantityType { type as! HKQuantityType }
```

LucasAssisRo commented on v2/RNHealthKitCore/HealthKitCore/HealthKitTypes.swift:
```diff
@@ -0,0 +1,143 @@
+        }
+    }
+
+    case BasalBodyTemperature
```
nitpick: 
```suggestion
    case basalBodyTemperature = "BasalBodyTemperature"
```

LucasAssisRo replied:
on second thought it is better the way you have it XD less typo chances

GGGava committed:
- Created QueryParameters superclass to remove code duplication
- Fixed typo

LucasAssisRo commented on v2/RNHealthKitCore/ReactNativeBridge/RNHealthKitWrapper.swift:
```diff
@@ -0,0 +1,121 @@
+    var core: HealthKitCore?
+
+    @objc
+    func initHealthKit(_ read: Array<String>, write: Array<String>, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
```
potential-issue: looks like is not possible to request `workout` permissions since this won't try to create it. I think you might have to go for the imperative code approach here
```swift
var readTypes: [any HealthKitType] = []
for identifier in read {
 guard let readType = QuantityType(rawValue: identifier) ?? WorkoutType(rawValue: identifier) else { continue }
 readTypes.append(readType)
}
```

GGGava replied:
Workout is not properly exported yet, currently working on that 😅

LucasAssisRo commented on v2/RNHealthKitCore/ReactNativeBridge/RNHealthKitWrapper.swift:
```diff
@@ -0,0 +1,121 @@
+    var core: HealthKitCore?
+
+    @objc
+    func initHealthKit(_ read: Array<String>, write: Array<String>, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
```
nitpick:


nitpick: 
```suggestion
    func initHealthKit(_ read: [String], write: [String], resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
```

LucasAssisRo reviewed (commented):
Looks good so far :D only had a few questions and nitpicks
Also found a potential issue, not sure if already on your radar

GGGava commented on v2/RNHealthKitCore/HealthKitCore/HealthKitTypes.swift:
```diff
@@ -0,0 +1,143 @@
+import Foundation
+import HealthKit
+
+public protocol HealthKitType: RawRepresentable {
```
Yes, nice, will do

GGGava commented on v2/RNHealthKitCore/HealthKitCore/Quantity/QuantityQueriesParameters.swift:
```diff
@@ -0,0 +1,151 @@
+public struct QuantityQuery {
+    let startDate: Date?
+    let endDate: Date?
+    let isUserEntered: Bool?
```
Samples manually added by the user on Health app. I'll think in a better name 😅

LucasAssisRo replied:
I see, but what's the difference between `false` and `nil` here?

GGGava replied:
Oh sorry lol

if set to false, it will only get samples that were **not** manually added. If `nil` it won't matter.

LucasAssisRo replied:
ohh I see makes sense now, my instinct on those situations is to go with an `OptionSet`

```swift
struct FetchMode: OptionSet {
    let rawValue: Int
    static var userEntered: Self { .init(rawValue: 1 << 0) }
    static var automaticallyEntered: Self { .init(rawValue: 1 << 1) }
}
```

then you can use it like `QuantityQuery(fetchMode: .userEntered)` or ` QuantityQuery(fetchMode: [.userEntered, .automaticallyEntered])`

`Bool?` is totally fine as well, in fact is probably a better choice if you plan on exposing this struct to objc

GGGava committed:
- Exported getWorkout method
- Fixed typo on statistical query
- Addded config to fix error warning on exmaple app
- Exported ids filter for quantity query

GGGava committed:
- Added save workout method
- Added missing resolve on save quantity query

thalysviana committed:
Add docs to HealthKitCore

thalysviana committed:
Update StatisticsQuantityQuery class description

thalysviana committed:
WIP

thalysviana committed:
Merge pull request #348 from agencyenterprise/chore/add-documentation-to-health-kit-core

chore: Add documentation to HealthKitCore module

thalysviana committed:
Process workout metadata

thalysviana committed:
Move WorkoutMetadata to index.ts

thalysviana committed:
Set deployment target to 14.0

thalysviana committed:
Refactor workout creation to use new workout api

thalysviana committed:
Remove automatic code signing

thalysviana committed:
Remove files

thalysviana committed:
Remove yarn.lock

thalysviana committed:
Comment keys with missing docs

thalysviana committed:
Merge pull request #351 from agencyenterprise/feature/add-metadata-to-workout-samples

feat: Add metadata to workout samples

thalysviana committed:
Add workout activities to sample

thalysviana committed:
Update example app with activities

thalysviana committed:
Update saveWorkout to include workout activities

thalysviana committed:
Fix issue with workout activitiy type assigned numbers

thalysviana committed:
Remove comments

thalysviana committed:
Update enum cases

thalysviana committed:
Update documentation

thalysviana committed:
Fix build due to missing dictionary casting

LucasReinaldo commented:
hey folks, I was just wondering if this is the current progress? https://github.com/agencyenterprise/react-native-health/projects/2

GGGava committed:
Merge pull request #352 from agencyenterprise/feature/add-actitivies-info-to-workout-samples

feat: Add activities info to workout samples

araadt commented:
Is V2 abandoned? I see a number of recent PRs for v1 haven't been merged. I'm curious as to the status of this project as a whole.

ruan-azevedo commented:
For everyone curious about the v2, it's not abandoned, but I do not have so much band to work in that lately.

The last status is that I tested this branch and it was broken in the last versions of React Native due to the RN new architecture, so we have to finish the refactoring for the bridge between React and the native code. Hopefully, we'll have this launched by the end of this year.

araadt commented:
I appreciate the update, @ruan-azevedo! I apologize—I didn't hope for that to sound accusatory. We've all got day-jobs and life away from the keyboard. I appreciate the efforts you're putting in!

ruan-azevedo commented:
Hey guys, I just want to let you know that I'm not working at AE Studio anymore and I'm not planning to continue working on this project.

If you want some feature, PR merged, or want to ask for someone to take this PR, please reach out to AE Studio guys on LinkedIn. I would suggest to reach out to Melanie or Felipe Taboada.

samuthekid commented:
First of all, thanks for this library! I appreciate a lot all the time and work you guys have put into it <3

How is the V2 going? I was really hoping to start the migration to the new RN Arch early next year, but this library is preventing me from doing that.
Do you guys have an ETA? Is this still being developed?

I think we would rather get the (bad?) news asap so we can plan a migration to another library :c

Thanks again for your work 🙏
